7_drivers_munge/src/GCM_driver_utils.R

```diff
@@ -242,26 +245,125 @@ download_gcm_data <- function(out_file_template, query_geom,
 
   # Build query_url
-  query_url <- sprintf("https://cida.usgs.gov/thredds/dodsC/notaro_%s_1980_1999", gcm_name)
-
-  # construct and submit query
-  gcm_job <- geoknife(
-    stencil = query_simplegeom,
-    fabric = webdata(
-      url = query_url,
-      variables = query_vars,
-      times = query_dates
+  query_url <- sprintf("https://cida.usgs.gov/thredds/dodsC/notaro_%s_%s",
+                       gcm_name, gcm_projection_period)
+
+  # Retry the GDP query multiple times if it doesn't work.
+  retry({
+    # construct and submit query
+    gcm_job <- geoknife(
+      stencil = query_simplegeom,
+      fabric = webdata(
+        url = query_url,
+        variables = query_vars,
+        times = query_dates
+      )
+      # The default knife algorithm is appropriate
     )
-    #TODO: should we specify the `knife`??
-  )
-  wait(gcm_job)
+    wait(gcm_job)
+    gcm_job_status <- check(gcm_job)$statusType
+  },
+  # Check the value of the last thing in the expr above (denoted
+  # by the ".") to decide if you should retry or not
+  until = ~ . == "ProcessSucceeded",
+  max_tries = 5)
+
   my_data <- result(gcm_job, with.units = TRUE)
 
   # Build out_file name
-  out_file <- sprintf(out_file_template, gcm_name, query_geom$tile_no[1])
+  out_file <- sprintf(out_file_template,
+                      gcm_name,
+                      gcm_projection_period,
+                      query_geom$tile_no[1])
 
   # write file
   arrow::write_feather(my_data, out_file)
 
   return(out_file)
 }
+
+#' @title Pull out individual file information from a dynamically mapped target
+#' @description Use a dynamically mapped target to extract information about
+#' the individual branch files to build a table that can group files and then
+#' trigger rebuilds when individual files within a group change.
+#' @param dynamic_branch_names character vector of the names of each branch created
+#' from running `names()` using the target dynamically branched target as input.
+#' @return a tibble with three columns (`name` = target name of the branch, `path`
+#' = character string of the file created by that branch, and `data` = hash code
+#' indicating the current state of the file) and a row for each branch.
+build_branch_file_hash_table <- function(dynamic_branch_names) {
+  # Get metadata for all the branches of the dynamic target
+  tar_meta(all_of(dynamic_branch_names), c("path", "data")) %>%
+    # There is just one path per branch, so remove the `list` format from this column
+    unnest(path)
+}
+
+#' @title Summarize hourly output from geoknife to daily
+#' @description the final GCM driver data will need to be daily, but geoknife
+#' returns hourly values. This step summarizes the data into daily values. It
+#' creates a file with the exact same name, except that the "_raw" part of the
+#' `in_file` filepath is replaced with "_daily".
+#' @param in_file filepath to a feather file containing the hourly geoknife data
+munge_to_glm <- function(in_file) {
```

Comment:
might want to call this `munge_notaro_to_glm` since this function is specific to a particular set of variables and unites. 

In the future, might want to pull out the generic munging components into a separate function and keep the munge prep work that is specific to a particular source in a unique function. (e.g., could use the same generic for NLDAS and Notaro GCMs, but unique prep functions for both). But probably not important now.

Reply:
Hmmm what would you say are the "generic munging components" here? Feels like most are Notaro specific

Reply:
I see generic as: going from hourly to daily, writing the files, and perhaps some of the variable changes. Probably good to just ignore this since I agree there isn't a clear divide and NLDAS isn't handled here currently anyhow.